src/inits.jl

```diff
@@ -0,0 +1,231 @@
+
+function _default_rng()
+    @static if VERSION >= v"1.7"
+        return Xoshiro(1234)
+    else
+        return MersenneTwister(1234)
+    end
+end
+
+"""
+    zeros32(rng::AbstractRNG, size...) = zeros(Float32, size...)
+
+Return an `Array{Float32}` of zeros of the given `size`. (`rng` is ignored)
+"""
+zeros32(rng::AbstractRNG, dims...) = zeros(rng, Float32, dims...)
+zeros32(dims...) = zeros32(_default_rng(), dims...)
+Base.zeros(rng::AbstractRNG, dims...) = zeros(dims...)
+"""
+    ones32(rng::AbstractRNG, size...) = ones(Float32, size...)
+
+Return an `Array{Float32}` of ones of the given `size`. (`rng` is ignored)
+"""
+ones32(rng::AbstractRNG, dims...) = ones(rng, Float32, dims...)
+ones32(dims...) = ones32(_default_rng(), dims...)
+Base.ones(rng::AbstractRNG, dims...) = ones(dims...)
+
+"""
+    randn32(rng::AbstractRNG, size...) = randn(rng, Float32, size...)
+
+Return an `Array{Float32}` of random numbers from a standard normal distribution of the
+given `size`.
+"""
+randn32(rng::AbstractRNG, dims...) = randn(rng, Float32, dims...)
+randn32(dims...) = randn32(_default_rng(), dims...)
+randn32(rng::AbstractRNG) = (rng, dims...) -> randn32(rng, dims...)
+randn32() = (dims...,) -> randn32(_default_rng(), dims...)
+
+"""
+    rand32(rng::AbstractRNG, size...) = rand(rng, Float32, size...)
+
+Return an `Array{Float32}` of random numbers from a uniform distribution of the given
+`size`.
+"""
+rand32(rng::AbstractRNG, dims...) = rand(rng, Float32, dims...)
+rand32(dims...) = rand32(_default_rng(), dims...)
+rand32(rng::AbstractRNG) = (rng, dims...) -> rand32(rng, dims...)
+rand32() = (dims...,) -> rand32(_default_rng(), dims...)
+
+"""
+    glorot_uniform(rng::AbstractRNG, size...; gain = 1)
+
+Return an `Array{Float32}` of the given `size` containing random numbers drawn from a
+uniform distribution on the interval ``[-x, x]``, where
+`x = gain * sqrt(6 / (fan_in + fan_out))`. This method is described in [1] and also known as
+Xavier initialization.
+
+# References
+
+[1] Glorot, Xavier, and Yoshua Bengio. "Understanding the difficulty of training deep
+feedforward neural networks." _Proceedings of the thirteenth international conference on
+artificial intelligence and statistics_. 2010.
+"""
+function glorot_uniform(rng::AbstractRNG, dims::Integer...; gain::Real=1)
+    scale = Float32(gain) * sqrt(24.0f0 / sum(_nfan(dims...)))
+    return (rand(rng, Float32, dims...) .- 0.5f0) .* scale
+end
+
+function glorot_uniform(dims::Integer...; kwargs...)
+    return glorot_uniform(_default_rng(), dims...; kwargs...)
+end
+
+function glorot_uniform(rng::AbstractRNG; init_kwargs...)
+    return (rng, dims...; kwargs...) -> glorot_uniform(rng,
+        dims...;
+        init_kwargs...,
+        kwargs...)
+end
+
+function glorot_uniform(; init_kwargs...)
+    return (dims...; kwargs...) -> glorot_uniform(_default_rng(),
+        dims...;
+        init_kwargs...,
+        kwargs...)
+end
```

Comment:
These functions can be combined using `rng=_default_rng()`. Also, let's use https://github.com/archermarx/PartialFunctions.jl so that the returned function is printed clearly rather than like a closure.

Reply:
> These functions can be combined using rng=_default_rng() 

I actually have them separated in order to have the closure for both calls of the form `init(dims...)` and `init(rng, dims...)`. If we default the rng the closure will still have the `init(rng, dims...)` signature. It shouldn't be a problem, but I figured we could offer both with the added definition in order to be as general as possible.

I'll rewrite the return functions using PartialFunctions, thanks for the suggestion!